contracts/TokensManager.sol

```diff
@@ -0,0 +1,67 @@
+pragma solidity ^0.6.0;
+import "@openzeppelin/contracts-upgradeable/proxy/Initializable.sol";
+import "@openzeppelin/contracts-upgradeable/access/AccessControlUpgradeable.sol";
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "abdk-libraries-solidity/ABDKMath64x64.sol";
+import "./cryptoblades.sol";
+
+contract TokensManager is Initializable, AccessControlUpgradeable {
+    using SafeMath for uint256;
+    using ABDKMath64x64 for int128;
+
+    bytes32 public constant GAME_ADMIN = keccak256("GAME_ADMIN");
+
+    CryptoBlades public game;
+
+    uint256 public combatTokenChargePercent;
+    // Note: All prices are multiplied by 100
+    uint256 public tokenPrice;
+    uint256 public skillTokenPrice;
+
+    modifier restricted() {
+        _restricted();
+        _;
+    }
+
+    function _restricted() internal view {
+        require(hasRole(GAME_ADMIN, msg.sender));
+    }
+
+    function initialize(
+        address gameContract
+    ) public initializer {
+        __AccessControl_init_unchained();
+        _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+
+        game = CryptoBlades(gameContract);
+        combatTokenChargePercent = 25;
+    }
+
+    function fight(uint256 char, uint256 wep, uint32 target, uint8 fightMultiplier) external payable {
+        uint256 expectedTokens = game.fight(char, wep, target, fightMultiplier);
+
+        uint256 offset = ABDKMath64x64.mulu(getSkillToNativeRatio(), expectedTokens.mul(combatTokenChargePercent));
+
+        require(msg.value == offset, 'Offset error');
+    }
+
+    function retrieve(address addressToTransferTo, uint256 amount) external restricted {
+        payable(addressToTransferTo).transfer(amount);
+    }
+
+    function getSkillToNativeRatio() public view returns (int128) {
+        return ABDKMath64x64.divu(skillTokenPrice, tokenPrice);
+    }
+
+    function setTokenPrice(uint256 price) external restricted {
+        tokenPrice = price;
+    }
+
```

Comment:
A little sidebar here, not necessary to change anything now but, for the future:
The bot that updates the native token and skill token prices might want to send updates for both at once, in which case it would make sense for a function that takes parameters for both. This is only really relevant for cutting costs on expensive networks, and only if the rate of updates is frequent enough to make a difference.